src/reports_updater.rs

```diff
@@ -1,53 +1,114 @@
-impl TcnMatcherImpl {
-
-  // TODO use TCN repo's match_btreeset test code? Compare performance.
   pub fn match_reports_with(tcns: Vec<ObservedTcn>, reports: Vec<SignedReport>) -> Result<Vec<MatchedReport>, ServicesError> {
-    let mut out: Vec<MatchedReport> = Vec::new();
-
     let observed_tcns_map: HashMap<[u8; 16], ObservedTcn> = tcns.into_iter().map(|e|
       (e.tcn.0, e)
     ).collect();
 
-    for report in reports {
-      // TODO no unwrap
-      let rep = report.clone().verify().unwrap();
-      for tcn in rep.temporary_contact_numbers() {
-        if let Some(entry) = observed_tcns_map.get(&tcn.0) {
-          out.push(MatchedReport { report, contact_time: entry.time.clone() });
-          break;
-        }
+    let observed_tcns_map = Arc::new(observed_tcns_map);
+
+    let res: Vec<Option<MatchedReport>> = reports.par_iter().map(|report| {
+      Self::match_report_with(&observed_tcns_map, report) 
+    }).collect();
+    
+
+    let res: Vec<MatchedReport> = res
+      .into_iter()
+      .filter_map(|option| option) // drop None (reports that didn't match)
+      .collect();
+
+    Ok(res)
+  }
+
+  pub fn match_report_with(observed_tcns_map: &HashMap<[u8; 16], ObservedTcn>, report: &SignedReport) -> Option<MatchedReport> {
+    let mut out: Option<MatchedReport> = None;
+
+    // TODO no unwrap
+    let rep = report.clone().verify().unwrap();
+    for tcn in rep.temporary_contact_numbers() {
+      if let Some(entry) = observed_tcns_map.get(&tcn.0) {
+        out = Some(MatchedReport { report: report.clone(), contact_time: entry.time.clone() });
+        break;
       }
     }
+    out
+  }
+}
 
-    Ok(out)
+pub struct TcnMatcherStdThreadSpawn {}
+
+impl TcnMatcher for TcnMatcherStdThreadSpawn {
+  fn match_reports(&self, tcns: Vec<ObservedTcn>, reports: Vec<SignedReport>) -> Result<Vec<MatchedReport>, ServicesError> {
+    Self::match_reports_with(tcns, reports)
+  }
+}
+ 
+impl TcnMatcherStdThreadSpawn {
+
+  pub fn match_reports_with(tcns: Vec<ObservedTcn>, reports: Vec<SignedReport>) -> Result<Vec<MatchedReport>, ServicesError> {
+    let observed_tcns_map: HashMap<[u8; 16], ObservedTcn> = tcns.into_iter().map(|e|
+      (e.tcn.0, e)
+    ).collect();
+
+    let observed_tcns_map = Arc::new(observed_tcns_map);
+
+    let threads: Vec<JoinHandle<Option<MatchedReport>>> = reports.into_iter().map(|report| {
+      let observed_tcns_map = observed_tcns_map.clone();
+      thread::spawn(move || Self::match_report_with(&observed_tcns_map, report)
+    )}).collect();
+
+    let res: Vec<MatchedReport> = threads
+      .into_iter()
+      .map(|t| t.join().unwrap())
+      .filter_map(|option| option) // drop None (reports that didn't match)
+      .collect();
+
+    Ok(res)
+  }
+
+  pub fn match_report_with(observed_tcns_map: &HashMap<[u8; 16], ObservedTcn>, report: SignedReport) -> Option<MatchedReport> {
+    let mut out: Option<MatchedReport> = None;
+
+    // TODO no unwrap
```

Comment:
"if let" should do the trick
https://doc.rust-lang.org/book/ch06-03-if-let.html

Reply:
This is more about how to handle errors. With `if let` we'd just ignore them.

Reply:
I created an issue https://github.com/Co-Epi/app-backend-rust/issues/38 to track this. We're using `unwrap` in several places.